#2749 Added phase to force Meta Data Source recalculation for Change Context

Patrykb0802 opened the pull request:
Added:
- Context point recalculation when enabling meta data source (execute())

Patrykb0802 committed:
#2749 Added phase to force Meta Data Source recalculation for Change Context

Added:
- Context point recalculation when enabling meta data source (execute())

Limraj commented on src/com/serotonin/mango/rt/dataSource/meta/MetaPointLocatorRT.java:
```diff
@@ -98,6 +98,7 @@ public void initialize(AbstractTimer timer, MetaDataSourceRT dataSource, DataPoi
 
         createContext();
 
+        execute(dataPoint.getPointValue());
```
1. How does it work when turning the point on and off, does it generate a duplicate value each time?
2. Is the dataPoint here always non-null?
3. Can 'getPointValue' method return null?
4. Is it safe to call the execute method at this place?
5. Recalculate should only be implemented for 'Change Context' type

Patrykb0802 committed:
How does it work when turning the point on and off, does it generate a duplicate value each time?
- It does not generate duplicate values each time

Is the dataPoint here always non-null?
Can 'getPointValue' method return null?

- It's not always not-null and  'getPointValue' can return null, that's why I added if condition checking if dataPoint.getPointValue() is null

Is it safe to call the execute method at this place?

- When execute was before rm.addDataPointListener() I was not sure, now it should be safe

Recalculate should only be implemented for 'Change Context' type

- Same if that checks if value is null is now checking update event type for context change update type before using execute()

Patrykb0802 committed:
Merge remote-tracking branch 'origin/release/2.7.7' into fix/#2749_Added_phase_to_force_Meta_Data_Source_recalculation_for_Change_Context

Patrykb0802 committed:
#2749 Added phase to force Meta Data Source recalculation for Change Context

Added condition which checks if point value changed while point was disabled

Patrykb0802 committed:
#2749 Added phase to force Meta Data Source recalculation for Change Context

- Instead of using execute(PointValueTime newValue, boolean initializeMode) we now use execute(long runtime, List<Integer> sourceIds, boolean initializeMode) for refreshing value when initializing point
- fixed if condition, instead of comparing pointValues now compares actual values

Patrykb0802 committed:
#2749 Added phase to force Meta Data Source recalculation for Change Context

- comparison is made by using a ObjectUtils.isEqual() method now

Patrykb0802 committed:
#2749 Added phase to force Meta Data Source recalculation for Change Context

- Added if condition checking for correct update event
- Added MetaPointLocatorRTTest.java which tests if certain update events trigger right methods
